# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [8]:
# import numpy and pandas
import numpy as np
import pandas as pd
from scipy.stats import ttest_rel, ttest_1samp, ttest_ind,chi2_contingency

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [3]:
# Run this code:

pokemon = pd.read_csv('../pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [3]:
# Your code here:
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [4]:
# Your code here:
pokemon.groupby("Legendary")['Legendary'].count()

Legendary
False    735
True      65
Name: Legendary, dtype: int64

There are 735 non-legendary pokemons and 65 legendary pokemons

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [5]:
# Your code here:
pokemon_legendary = pokemon[pokemon['Legendary']==True]
pokemon_nonlegendary = pokemon[pokemon['Legendary']==False]
display(pokemon_legendary.describe())
display(pokemon_nonlegendary.describe())

#       Total          HP      Attack     Defense     Sp. Atk  \
count   65.000000   65.000000   65.000000   65.000000   65.000000   65.000000   
mean   470.215385  637.384615   92.738462  116.676923   99.661538  122.184615   
std    173.651095   60.937389   21.722164   30.348037   28.255131   31.104608   
min    144.000000  580.000000   50.000000   50.000000   20.000000   50.000000   
25%    381.000000  580.000000   80.000000  100.000000   90.000000  100.000000   
50%    483.000000  600.000000   91.000000  110.000000  100.000000  120.000000   
75%    642.000000  680.000000  105.000000  131.000000  115.000000  150.000000   
max    721.000000  780.000000  150.000000  190.000000  200.000000  194.000000   

          Sp. Def       Speed  Generation  
count   65.000000   65.000000   65.000000  
mean   105.938462  100.184615    3.769231  
std     28.827004   22.952323    1.455262  
min     20.000000   50.000000    1.000000  
25%     90.000000   90.000000    3.000000  
50%    100.000000  100.000000    4.000000  
75%    120.000000  110.000000    5.000000  
max    200.000000  180.000000    6.000000

#       Total          HP      Attack     Defense     Sp. Atk  \
count  735.000000  735.000000  735.000000  735.000000  735.000000  735.000000   
mean   353.315646  417.213605   67.182313   75.669388   71.559184   68.454422   
std    208.590419  106.760417   24.808849   30.490153   30.408194   29.091705   
min      1.000000  180.000000    1.000000    5.000000    5.000000   10.000000   
25%    175.500000  324.000000   50.000000   54.500000   50.000000   45.000000   
50%    346.000000  425.000000   65.000000   72.000000   66.000000   65.000000   
75%    533.500000  498.000000   79.500000   95.000000   85.000000   85.000000   
max    715.000000  700.000000  255.000000  185.000000  230.000000  175.000000   

          Sp. Def       Speed  Generation  
count  735.000000  735.000000  735.000000  
mean    68.892517   65.455782    3.284354  
std     25.669310   27.843038    1.673471  
min     20.000000    5.000000    1.000000  
25%     50.000000   45.000000    2.000000  
50%     65.000000   64.000000    3.000000  
75%     85.000000   85.000000    5.000000  
max    230.000000  160.000000    6.000000

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [6]:
# Your code here:
ttest = ttest_ind(pokemon_legendary['Total'],pokemon_nonlegendary['Total'], equal_var=False)
ttest

Ttest_indResult(statistic=25.8335743895517, pvalue=9.357954335957446e-47)

What do you conclude from this test? Write your conclusions below.

In [7]:
# Your conclusions here:
print(f'The p-value is: {ttest.pvalue}. It is lower than 0.05, so significant differences between both comparing total points were observed')


The p-value is: 9.357954335957446e-47. It is lower than 0.05, so significant differences between both comparing total points were observed


How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [8]:
# Your code here:
pokemon.groupby("Type 1")["Type 1"].count()

Type 1
Bug          69
Dark         31
Dragon       32
Electric     44
Fairy        17
Fighting     27
Fire         52
Flying        4
Ghost        32
Grass        70
Ground       32
Ice          24
Normal       98
Poison       28
Psychic      57
Rock         44
Steel        27
Water       112
Name: Type 1, dtype: int64

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [9]:
# Your code here:
water= pokemon["Type 1"]=="Water"
nonwater=  pokemon["Type 1"]!="Water"
display(pokemon[water].describe()[1:3])
display(pokemon[nonwater].describe()[1:3])


#       Total         HP     Attack    Defense    Sp. Atk  \
mean  303.089286  430.455357  72.062500  74.151786  72.946429  74.812500   
std   188.440807  113.188266  27.487026  28.377192  27.773809  29.030128   

        Sp. Def      Speed  Generation  
mean  70.517857  65.964286    2.857143  
std   28.460493  23.019353    1.558800

#       Total         HP     Attack    Defense    Sp. Atk  \
mean  372.536337  435.859012  68.802326  79.790698  73.988372  72.495640   
std   209.928799  121.091682  25.194299  33.025152  31.719933  33.292537   

        Sp. Def      Speed  Generation  
mean  72.127907  68.654070    3.399709  
std   27.739292  29.925907    1.666119

Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

In [10]:
# Your code here:
ttest = ttest_ind(pokemon[water]['Total'],pokemon[nonwater]['Total'], equal_var=True)
ttest

Ttest_indResult(statistic=-0.4418547448849676, pvalue=0.6587140317488793)

Write your conclusion below.

In [11]:
# Your conclusions here:
print(f'The p-value is: {ttest.pvalue}. It is higher than 0.05, so significant differences between both comparing total points were not observed')

The p-value is: 0.6587140317488793. It is higher than 0.05, so significant differences between both comparing total points were not observed


# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [12]:
# Your code here:
"""
H0 = Defense and attack are not equal
H1 = Defense and attack are equal
"""
ttest_rel = ttest_rel(pokemon['Attack'], pokemon['Defense'])
ttest_rel

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

Describe the results of the test in the cell below.

In [13]:
# Your conclusions here:
print(f'The p-value is: {ttest_rel.pvalue}. It is lower than 0.05, so Attack and Defense are equal')

The p-value is: 1.7140303479358558e-05. It is lower than 0.05, so Attack and Defense are equal


We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [5]:
# Your code here:
"""
H0 = SP. Defense and SP.attack are not equal
H1 = Sp.Defense and Sp.attack are equal
"""
ttest_rel_sp = ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def'])
ttest_rel_sp


Ttest_relResult(statistic=0.853986188453353, pvalue=0.3933685997548122)

Describe the results of the test in the cell below.

In [ ]:
# Your conclusions here:
print(f'The p-value is: {ttest_rel_sp.pvalue}. It is higher than 0.05, so Attack and Defense are not equal')


As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [7]:
# Your code here:
pokemon['diff'] = pokemon['Attack'] - pokemon['Defense']
tt_diff = ttest_1samp(pokemon['diff'], 0)  
print(tt_diff)
print(f'The p-value is: {tt_diff.pvalue}. It is lower than 0.05, so Attack and Defense are equal')

Ttest_1sampResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)
The p-value is: 1.7140303479358558e-05. It is lower than 0.05, so Attack and Defense are equal


# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [13]:
# Your code here:
pokemon["is_water"] = pokemon["Type 1"].apply(lambda x: True if x=="Water" else False)
contingency_table = pd.crosstab(pokemon['Legendary'], pokemon["is_water"] )
contingency_table

is_water   False  True 
Legendary              
False        627    108
True          61      4

Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [18]:
# Your code here:
chi2, p, dof, ex = chi2_contingency(contingency_table, correction=False)
print(chi2,p,dof,ex)

3.617455159932616 0.0571762700873503 1 [[632.1 102.9]
 [ 55.9   9.1]]


Based on a 95% confidence, should we reject the null hypothesis?

In [19]:
# Your answer here:
print(f'The p value is {p}, greater than 0.05, so we can not reject null hypothesis')


The p value is 0.0571762700873503, greater than 0.05, so we can not reject null hypothesis
